pkg/minishift/systemtray/systemtray.go

```diff
@@ -0,0 +1,248 @@
+*/
+
+package systemtray
+
+import (
+	"fmt"
+	goos "os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/anjannath/systray"
+	"github.com/docker/machine/libmachine"
+	libmachineState "github.com/docker/machine/libmachine/state"
+	"github.com/minishift/minishift/cmd/minishift/state"
+	"github.com/minishift/minishift/pkg/minikube/cluster"
+	"github.com/minishift/minishift/pkg/minishift/profile"
+	"github.com/minishift/minishift/pkg/util/os"
+	"github.com/minishift/minishift/pkg/util/slice"
+)
+
+var (
+	submenus            = make(map[string]*systray.MenuItem)
+	submenusToMenuItems = make(map[string]MenuAction)
+	iconStart, _        = filepath.Abs("running.bmp")
+	iconStop, _         = filepath.Abs("stopped.bmp")
+	iconDne, _          = filepath.Abs("doesnotexist.bmp")
+
+	profiles        []string
+	profileMenuList []*systray.MenuItem
+
+	submenusLock            sync.RWMutex
+	submenusToMenuItemsLock sync.RWMutex
+)
+
+type MenuAction struct {
+	start *systray.MenuItem
+	stop  *systray.MenuItem
+}
+
+func OnReady() {
+	iconpath, _ := filepath.Abs("minishift.ico")
+	systray.SetIconPath(iconpath)
+	exit := systray.AddMenuItem("Exit", "", 0)
+	systray.AddSeparator()
+	profiles = profile.GetProfileList()
+	for _, profile := range profiles {
+		submenu := systray.AddSubMenu(profile)
+		startMenu := submenu.AddSubMenuItem("Start", "", 0)
+		stopMenu := submenu.AddSubMenuItem("Stop", "", 0)
+		iconpath, _ := filepath.Abs("doesnotexist.bmp")
+		submenu.AddBitmap(iconpath)
+		submenus[profile] = submenu
+		submenusToMenuItems[profile] = MenuAction{start: startMenu, stop: stopMenu}
+	}
+
+	go func() {
+		<-exit.OnClickCh()
+		systray.Quit()
+	}()
+
+	for k, v := range submenusToMenuItems {
+		fmt.Println(k)
+		go func(iconpath, submenu string, v MenuAction) {
+			for {
+				<-v.start.OnClickCh()
+				err := startProfile(submenu)
+				if err == nil {
+					submenusLock.Lock()
+					submenus[submenu].AddBitmap(iconpath)
+					submenusLock.Unlock()
+				}
+			}
+		}(iconStart, k, v)
+
+		go func(iconpath, submenu string, v MenuAction) {
+			for {
+				<-v.stop.OnClickCh()
+				err := stopProfile(submenu)
+				if err == nil {
+					submenusLock.Lock()
+					submenus[submenu].AddBitmap(iconpath)
+					submenusLock.Unlock()
+				}
+			}
+		}(iconStop, k, v)
+	}
+
+	go addNewProfilesToTray()
+
+	go removeDeletedProfilesFromTray()
+
+	go updateProfileStatus(iconDne)
+}
+
+func OnExit() {
+	return
+}
+
+func getStatus(profileName string) int {
+	api := libmachine.NewClient(state.InstanceDirs.Home, state.InstanceDirs.Certs)
+	defer api.Close()
+
+	vmStatus, err := cluster.GetHostStatus(api, profileName)
+	if err != nil {
+		return 0
+	}
+
+	if vmStatus == libmachineState.Running.String() {
+		return 1
+	}
+	if vmStatus == libmachineState.Stopped.String() {
+		return 2
+	}
+	return 0
+}
+
+// Add newly created profiles to the tray
+func addNewProfilesToTray() {
+	for {
+		time.Sleep(30 * time.Second)
+
+		newProfilesList := profile.GetProfileList()
```

Comment:
duplicated logic, which feels can be shared between implementations that handle the actual list.